torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,293 @@
+//
+// `TensorDataContainer({{1}, {2}})`
+//
+// which matches to the `TensorDataContainer(std::initializer_list<TensorDataContainer>)` constructor,
+// and in an attempt to convert `{1}` and `{2}` to `TensorDataContainer`, it calls the following:
+//
+// `TensorDataContainer({1})`  (same call path happens for `{2}`, and we'll just focus on `{1}` here)
+//
+// At this point, theoretically there are two plausible ways for `{1}` to be matched to one of the
+// constructors of `TensorDataContainer`:
+//
+// 1. It can be a list-initialization of a scalar value, thus matching `TensorDataContainer(int value)`.
+// 2. It can be converted to `std::initializer_list<TensorDataContainer>`, thus matching
+//    `TensorDataContainer(std::initializer_list<TensorDataContainer>)`.
+//
+// How does the compiler decide which one to choose? According to
+// `https://en.cppreference.com/w/cpp/language/list_initialization`, braced-init-list always prefers
+// the constructor that takes `std::initializer_list`. Hence we happily move forward with constructor #2,
+// and it calls the following:
+//
+// `TensorDataContainer(1)`
+//
+// Now it matches `TensorDataContainer(int value)`, which stores `1` as a scalar value. All is good.
+struct TensorDataContainer {
+  // NOTE: For tensors with zero-size dimensions (e.g. `torch::tensor({{}, {}})`),
+  // the innermost empty braced-init-list `{}` matches the default constructor of
+  // the innermost `TensorDataContainer`.
+  TensorDataContainer() :
+      sizes_({0}),
+      scalar_type_(at::kFloat),
+      type_(TensorDataContainerType::InitList) {}
+#define TENSOR(T, S) \
+  TensorDataContainer(T value) : \
+      sizes_(), \
+      scalar_type_(at::k##S), \
+      type_(TensorDataContainerType::Scalar), \
+      scalar_(value) {}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+  TensorDataContainer(std::initializer_list<TensorDataContainer> init_list) :
+      sizes_(),
+      scalar_type_(init_list.begin()->scalar_type()),
+      type_(TensorDataContainerType::InitList),
+      init_list_(init_list) {
+    const TensorDataContainer& first_elem = *(init_list.begin());
+    for (const auto& elem : init_list) {
+      TORCH_CHECK(elem.sizes() == first_elem.sizes(),
+        "Expected all sub-lists to have sizes: ",
+        first_elem.sizes(),
+        " (e.g. ", first_elem, "), ",
+        "but got sub-list ",
+        elem,
+        " with sizes: ",
+        elem.sizes());
+      TORCH_CHECK(elem.scalar_type() == first_elem.scalar_type(),
+        "Expected all elements of the tensor to have the same scalar type: ",
+        first_elem.scalar_type(),
+        ", but got element of scalar type: ",
+        elem.scalar_type());
+    }
+    sizes_.reserve(first_elem.sizes().size() + 1);
+    sizes_.push_back(init_list.size());
+    sizes_.insert(sizes_.end(), first_elem.sizes().begin(), first_elem.sizes().end());
+  }
+
+#define TENSOR(T, S) \
+  TensorDataContainer(at::ArrayRef<T> values) : \
```

Comment:
Just checking, we don't (and we didn't) support `torch::tensor(tensor)` before, right?

Reply:
Right we didn't support it, and we currently don't support it either.